src/app/clusters/operational-credentials-server/operational-credentials-server.cpp

```diff
@@ -360,6 +362,182 @@ bool emberAfOperationalCredentialsClusterAddOpCertCallback(chip::app::Command *
+    SuccessOrExit(err = tlvWriter.Put(TLV::ContextTag(1), ByteSpan(nullptr, 0))); // TODO: add firmware digest(s)
+    SuccessOrExit(err = tlvWriter.EndContainer(outerContainerType));
+    SuccessOrExit(err = tlvWriter.Finalize(&firmwareInfo));
+
+    attestationElements = System::PacketBufferHandle::New(System::PacketBuffer::kMaxSize);
+    VerifyOrExit(!attestationElements.IsNull(), err = CHIP_ERROR_NO_MEMORY);
+
+    attestationElementsTbs = System::PacketBufferHandle::New(System::PacketBuffer::kMaxSize);
+    VerifyOrExit(!attestationElementsTbs.IsNull(), err = CHIP_ERROR_NO_MEMORY);
+
+    tlvWriter.Init(std::move(attestationElements));
+    outerContainerType = TLV::kTLVType_NotSpecified;
+    SuccessOrExit(err = tlvWriter.StartContainer(TLV::AnonymousTag, TLV::kTLVType_Structure, outerContainerType));
+    SuccessOrExit(err = tlvWriter.Put(TLV::ContextTag(0), ByteSpan(certDeclaration->Start(), certDeclaration->DataLength())));
+    SuccessOrExit(err = tlvWriter.Put(TLV::ContextTag(1), attestationNonce));
+    SuccessOrExit(err = tlvWriter.Put(TLV::ContextTag(2), ByteSpan(nullptr, 0))); // TODO: add timestamp
+    SuccessOrExit(err = tlvWriter.Put(TLV::ContextTag(3), ByteSpan(firmwareInfo->Start(), firmwareInfo->DataLength())));
+    SuccessOrExit(err = tlvWriter.Put(TLV::ContextTag(4), ByteSpan(nullptr, 0))); // Vendor Reserved 5
+    SuccessOrExit(err = tlvWriter.Put(TLV::ContextTag(5), ByteSpan(nullptr, 0))); // Vendor Reserved 6
+    SuccessOrExit(err = tlvWriter.Put(TLV::ContextTag(6), ByteSpan(nullptr, 0))); // Vendor Reserved 7
+    SuccessOrExit(err = tlvWriter.Put(TLV::ContextTag(7), ByteSpan(nullptr, 0))); // Vendor Reserved 8
+    SuccessOrExit(err = tlvWriter.EndContainer(outerContainerType));
+    SuccessOrExit(err = tlvWriter.Finalize(&attestationElements));
+
+    // Retrieve attestation challenge
+    // TODO: Retrieve Peer Node Id / Remove hard-coded NodeId value
+    state = GetGlobalSecureSessionMgr().GetPeerConnectionState({ 112233, kAnyKeyId, kUndefinedAdminId });
+    VerifyOrExit(state != nullptr, err = CHIP_ERROR_NOT_CONNECTED);
+    attestationChallenge = state->GetSecureSession().GetAttestationChallenge();
+
+    tlvWriter.Init(std::move(attestationElementsTbs));
+    outerContainerType = TLV::kTLVType_NotSpecified;
+    SuccessOrExit(err = tlvWriter.StartContainer(TLV::AnonymousTag, TLV::kTLVType_Structure, outerContainerType));
+    SuccessOrExit(err =
+                      tlvWriter.Put(TLV::ContextTag(0), ByteSpan(attestationElements->Start(), attestationElements->DataLength())));
+    SuccessOrExit(err = tlvWriter.Put(TLV::ContextTag(1), attestationChallenge));
+    SuccessOrExit(err = tlvWriter.EndContainer(outerContainerType));
+    SuccessOrExit(err = tlvWriter.Finalize(&attestationElementsTbs));
+
+    SuccessOrExit(err = keypair.ECDSA_sign_msg(attestationElementsTbs->Start(), attestationElementsTbs->DataLength(), signature));
+
+    VerifyOrExit(commandObj != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
```

Comment:
This check must be at the very beginning. If we have invalid commandObj, we should not pay the cost of ECDSA signature which is very costly on some embedded platforms.

Reply:
udated: moved to beginning